infrastructure/smart-contracts/contracts/upgrades/esXai/esXai4.sol

```diff
@@ -0,0 +1,380 @@
+     * @return The address of the whitelisted account.
+     */
+    function getWhitelistedAddressAtIndex(uint256 index) public view returns (address) {
+        require(index < getWhitelistCount(), "Index out of bounds");
+        return _whitelist.at(index);
+    }
+
+    /**
+     * @dev Function to get the count of whitelisted addresses.
+     * @return The count of whitelisted addresses.
+     */
+    function getWhitelistCount() public view returns (uint256) {
+        return _whitelist.length();
+    }
+
+    /**
+     * @dev Override the transfer function to only allow addresses that are in the white list in the to or from field to go through
+     * @param to The address to transfer to.
+     * @param amount The amount to transfer.
+     */
+    function transfer(address to, uint256 amount) public override returns (bool) {
+        require(_whitelist.contains(msg.sender) || _whitelist.contains(to), "Transfer not allowed: address not in whitelist");
+        return super.transfer(to, amount);
+    }
+
+    /**
+     * @dev Override the transferFrom function to only allow addresses that are in the white list in the to or from field to go through
+     * @param from The address to transfer from.
+     * @param to The address to transfer to.
+     * @param amount The amount to transfer.
+     */
+    function transferFrom(address from, address to, uint256 amount) public override returns (bool) {
+        require(_whitelist.contains(from) || _whitelist.contains(to), "Transfer not allowed: address not in whitelist");
+        return super.transferFrom(from, to, amount);
+    }
+
+    /**
+     * @dev Function to start the redemption process
+     * @param amount The amount of esXai to redeem.
+     * @param duration The duration of the redemption process in seconds.
+     */
+    function startRedemption(uint256 amount, uint256 duration) public {
+        require(_redemptionActive, "Redemption is currently inactive");
+        require(amount > 0, "Invalid Amount");
+        require(balanceOf(msg.sender) >= amount, "Insufficient esXai balance");
+        require(duration == 15 days || duration == 90 days || duration == 180 days, "Invalid duration");
+        
+        // Check if the sender failed KYC
+        bool failedKyc = PoolFactory3(poolFactoryAddress).failedKyc(msg.sender);
+        require(!failedKyc, "KYC failed, cannot redeem");
+        
+        // No longer transferring esXai from the sender's account to this contract
+        // Stoing redemption claim as voucherIssued instead
+
+        // Confirm the user has the appropriate amount of esXai available
+        uint256 currentBalance = balanceOf(msg.sender);        
+        uint256 totalesXaiStaked = PoolFactory3(poolFactoryAddress).getTotalesXaiStakedByUser(msg.sender);
+        uint256 totalesXaiPendingRedemption = _totalPendingRedemptions[msg.sender];
+
+        uint256 availableesXai = currentBalance + totalesXaiStaked - totalesXaiPendingRedemption;
```

Comment:
Just a comment here, every variable costs gas on the EVM, this is not going to be any significant amount, however keep it in mind, it's generally a good practice to consider gas efficiency when writing EVM programs.

Reply:
`availableesXai`, `totalesXaiStaked`, `totalesXaiPendingRedemption` don't follow the camel case naming convention, is this on purpose ?

Reply:
The contract name is like that, I have corrected.

Reply:
>The contract name is cased like that...

Do you mean esXai starting with a small cased "e" ? That should still be a capital case E for camel casing right ?

Please have a look at the PoolFactory function, for me personally it gets very hard to read like that ^^

Reply:
I agree and will adopt moving forward.